// sparql-engine/dist/engine/stages/update-stage-builder.js
/* file : update-stage-builder.ts
MIT License

Copyright (c) 2018-2020 Thomas Minier

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var stage_builder_1 = require("./stage-builder");
var pipeline_1 = require("../pipeline/pipeline");
var consumer_1 = require("../../operators/update/consumer");
var insert_consumer_1 = require("../../operators/update/insert-consumer");
var delete_consumer_1 = require("../../operators/update/delete-consumer");
var clear_consumer_1 = require("../../operators/update/clear-consumer");
var many_consumers_1 = require("../../operators/update/many-consumers");
var construct_1 = require("../../operators/modifiers/construct");
var rewritings = require("./rewritings.js");
var bindings_1 = require("../../rdf/bindings");
var symbols_1 = require("../context/symbols");
var nop_consumer_1 = require("../../operators/update/nop-consumer");
var action_consumer_1 = require("../../operators/update/action-consumer");
/**
 * An UpdateStageBuilder evaluates SPARQL UPDATE queries.
 * @see https://www.w3.org/TR/2013/REC-sparql11-update-20130321
 * @author Thomas Minier
 */
var UpdateStageBuilder = /** @class */ (function (_super) {
    __extends(UpdateStageBuilder, _super);
    function UpdateStageBuilder() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Create a {@link Consumable} used to evaluate a SPARQL 1.1 Update query
     * @param updates - Set of Update queries to execute
     * @param options - Execution options
     * @return A Consumable used to evaluatethe set of update queries
     */
    UpdateStageBuilder.prototype.execute = function (updates, context) {
        var _this = this;
        var queries;
        return new many_consumers_1.default(updates.map(function (update) {
            if ('updateType' in update) {
                switch (update.updateType) {
                    case 'insert':
                    case 'delete':
                    case 'insertdelete':
                        return _this._handleInsertDelete(update, context);
                    default:
                        return new consumer_1.ErrorConsumable("Unsupported SPARQL UPDATE query: " + update.updateType);
                }
            }
            else if ('type' in update) {
                switch (update.type) {
                    case 'create': {
                        var createNode = update;
                        var iri_1 = createNode.graph.name;
                        if (_this._dataset.hasNamedGraph(iri_1)) {
                            if (!createNode.silent) {
                                return new consumer_1.ErrorConsumable("Cannot create the Graph with iri " + iri_1 + " as it already exists in the RDF dataset");
                            }
                            return new nop_consumer_1.default();
                        }
                        return new action_consumer_1.default(function () {
                            _this._dataset.addNamedGraph(iri_1, _this._dataset.createGraph(iri_1));
                        });
                    }
                    case 'drop': {
                        var dropNode = update;
                        // handle DROP DEFAULT queries
                        if ('default' in dropNode.graph && dropNode.graph.default) {
                            return new action_consumer_1.default(function () {
                                var defaultGraphIRI = _this._dataset.getDefaultGraph().iri;
                                if (_this._dataset.iris.length < 1) {
                                    return new consumer_1.ErrorConsumable("Cannot drop the default Graph with iri " + iri_2 + " as it would leaves the RDF dataset empty without a default graph");
                                }
                                var newDefaultGraphIRI = _this._dataset.iris.find(function (iri) { return iri !== defaultGraphIRI; });
                                _this._dataset.setDefaultGraph(_this._dataset.getNamedGraph(newDefaultGraphIRI));
                            });
                        }
                        // handle DROP ALL queries
                        if ('all' in dropNode.graph && dropNode.graph.all) {
                            return new action_consumer_1.default(function () {
                                _this._dataset.iris.forEach(function (iri) { return _this._dataset.deleteNamedGraph(iri); });
                            });
                        }
                        // handle DROP GRAPH queries
                        var iri_2 = dropNode.graph.name;
                        if (!_this._dataset.hasNamedGraph(iri_2)) {
                            if (!dropNode.silent) {
                                return new consumer_1.ErrorConsumable("Cannot drop the Graph with iri " + iri_2 + " as it doesn't exists in the RDF dataset");
                            }
                            return new nop_consumer_1.default();
                        }
                        return new action_consumer_1.default(function () {
                            _this._dataset.deleteNamedGraph(iri_2);
                        });
                    }
                    case 'clear':
                        return _this._handleClearQuery(update);
                    case 'add':
                        return _this._handleInsertDelete(rewritings.rewriteAdd(update, _this._dataset), context);
                    case 'copy':
                        // A COPY query is rewritten into a sequence [CLEAR query, INSERT query]
                        queries = rewritings.rewriteCopy(update, _this._dataset);
                        return new many_consumers_1.default([
                            _this._handleClearQuery(queries[0]),
                            _this._handleInsertDelete(queries[1], context)
                        ]);
                    case 'move':
                        // A MOVE query is rewritten into a sequence [CLEAR query, INSERT query, CLEAR query]
                        queries = rewritings.rewriteMove(update, _this._dataset);
                        return new many_consumers_1.default([
                            _this._handleClearQuery(queries[0]),
                            _this._handleInsertDelete(queries[1], context),
                            _this._handleClearQuery(queries[2])
                        ]);
                    default:
                        return new consumer_1.ErrorConsumable("Unsupported SPARQL UPDATE query: " + update.type);
                }
            }
            return new consumer_1.ErrorConsumable("Unsupported SPARQL UPDATE query: " + update);
        }));
    };
    /**
     * Build a Consumer to evaluate SPARQL UPDATE queries
     * @private
     * @param update  - Parsed query
     * @param options - Execution options
     * @return A Consumer used to evaluate SPARQL UPDATE queries
     */
    UpdateStageBuilder.prototype._handleInsertDelete = function (update, context) {
        var _this = this;
        var engine = pipeline_1.Pipeline.getInstance();
        var source = engine.of(new bindings_1.BindingBase());
        var graph = null;
        var consumables = [];
        if (update.updateType === 'insertdelete') {
            graph = ('graph' in update) ? this._dataset.getNamedGraph(update.graph) : null;
            // evaluate the WHERE clause as a classic SELECT query
            var node = {
                prefixes: context.getProperty(symbols_1.default.PREFIXES),
                type: 'query',
                where: update.where,
                queryType: 'SELECT',
                variables: ['*'],
                // copy the FROM clause from the original UPDATE query
                from: ('from' in update) ? update.from : undefined
            };
            source = this._builder._buildQueryPlan(node, context);
        }
        // clone the source first
        source = engine.clone(source);
        // build consumers to evaluate DELETE clauses
        if ('delete' in update && update.delete.length > 0) {
            consumables = consumables.concat(update.delete.map(function (v) {
                return _this._buildDeleteConsumer(source, v, graph, context);
            }));
        }
        // build consumers to evaluate INSERT clauses
        if ('insert' in update && update.insert.length > 0) {
            consumables = consumables.concat(update.insert.map(function (v) {
                return _this._buildInsertConsumer(source, v, graph, context);
            }));
        }
        return new many_consumers_1.default(consumables);
    };
    /**
     * Build a consumer to evaluate a SPARQL INSERT clause
     * @private
     * @param source - Input {@link PipelineStage}
     * @param group - parsed SPARQL INSERT clause
     * @param graph - RDF Graph used to insert data
     * @return A consumer used to evaluate a SPARQL INSERT clause
     */
    UpdateStageBuilder.prototype._buildInsertConsumer = function (source, group, graph, context) {
        var tripleSource = construct_1.default(source, { template: group.triples });
        if (graph === null) {
            graph = (group.type === 'graph' && 'name' in group) ? this._dataset.getNamedGraph(group.name) : this._dataset.getDefaultGraph();
        }
        return new insert_consumer_1.default(tripleSource, graph, context);
    };
    /**
     * Build a consumer to evaluate a SPARQL DELETE clause
     * @private
     * @param  source - Input {@link PipelineStage}
     * @param  group - parsed SPARQL DELETE clause
     * @param  graph - RDF Graph used to delete data
     * @return A consumer used to evaluate a SPARQL DELETE clause
     */
    UpdateStageBuilder.prototype._buildDeleteConsumer = function (source, group, graph, context) {
        var tripleSource = construct_1.default(source, { template: group.triples });
        if (graph === null) {
            graph = (group.type === 'graph' && 'name' in group) ? this._dataset.getNamedGraph(group.name) : this._dataset.getDefaultGraph();
        }
        return new delete_consumer_1.default(tripleSource, graph, context);
    };
    /**
     * Build a Consumer to evaluate CLEAR queries
     * @private
     * @param query - Parsed query
     * @return A Consumer used to evaluate CLEAR queries
     */
    UpdateStageBuilder.prototype._handleClearQuery = function (query) {
        var graph = null;
        var iris = this._dataset.iris;
        if (query.graph.default) {
            graph = this._dataset.getDefaultGraph();
        }
        else if (query.graph.all) {
            graph = this._dataset.getUnionGraph(iris, true);
        }
        else if (query.graph.named) {
            graph = this._dataset.getUnionGraph(iris, false);
        }
        else {
            graph = this._dataset.getNamedGraph(query.graph.name);
        }
        return new clear_consumer_1.default(graph);
    };
    return UpdateStageBuilder;
}(stage_builder_1.default));
exports.default = UpdateStageBuilder;
