// vaultifier/dist/main/environments/web.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VaultifierWeb = void 0;
const oidc_client_1 = require("oidc-client");
const __1 = require("..");
const constants_1 = require("../constants");
const crypto_1 = require("../crypto");
const interfaces_1 = require("../interfaces");
const storage_1 = require("../storage");
const urls_1 = require("../urls");
const defaultOptions = {
    repo: undefined,
    clientId: undefined,
    baseUrlParamName: 'PIA_URL',
    appKeyParamName: 'APP_KEY',
    appSecretParamName: 'APP_SECRET',
    masterKeyParamName: 'MASTER_KEY',
    nonceParamName: 'NONCE',
    clientIdParamName: 'client_id',
    clientSecretParamName: 'client_secret',
    authorizationCodeParamName: 'code',
    stateParamName: 'state',
};
class VaultifierWeb {
    constructor(options, vaultifier) {
        this.options = options;
        this.vaultifier = vaultifier;
        this.initialize = (config = {}) => __awaiter(this, void 0, void 0, function* () {
            var _a;
            const vaultifier = this.vaultifier;
            if (!vaultifier)
                return undefined;
            // vaultifier must be valid in order to proceed with initialization
            if (!vaultifier.isValid())
                return undefined;
            const { oAuthType, } = config;
            const { appKeyParamName, appSecretParamName, authorizationCodeParamName, stateParamName, clientIdParamName, clientSecretParamName, masterKeyParamName, nonceParamName, } = this.options;
            let { clientId, } = this.options;
            const getParam = VaultifierWeb._getParamAccessor();
            const appKey = getParam(appKeyParamName);
            const appSecret = getParam(appSecretParamName);
            // if clientId parameter is specified as query parameter it is already the second step client id parameter of OAuth
            clientId = (_a = getParam(clientIdParamName)) !== null && _a !== void 0 ? _a : clientId;
            const authorizationCode = getParam(authorizationCodeParamName);
            let credentials = undefined;
            if (appKey && appSecret)
                credentials = {
                    appKey,
                    appSecret,
                };
            else if (authorizationCode && clientId)
                credentials = {
                    authorizationCode,
                    clientId,
                    state: getParam(stateParamName),
                };
            const masterKey = getParam(masterKeyParamName);
            const nonce = getParam(nonceParamName);
            const end2end = (masterKey && nonce) ? {
                masterKey,
                nonce,
            } : undefined;
            vaultifier.credentials = credentials;
            vaultifier.privateKeyCredentials = end2end;
            try {
                // try initializing vaultifier to see if credentials are working
                yield vaultifier.initialize();
            }
            catch ( /* */_b) { /* */ }
            // if we could not authorize until this stage
            // we try to login via OAuth, if supported
            const isAuthenticated = yield vaultifier.isAuthenticated();
            if (!isAuthenticated) {
                const oAuthSupport = oAuthType;
                if (clientId && oAuthSupport && oAuthSupport.type === interfaces_1.OAuthType.AUTHORIZATION_CODE) {
                    // create PKCE secret
                    const pkceSecret = crypto_1.getRandomString(32);
                    // const hashedSecret = btoa(await createSha256Hex(pkceSecret));
                    const redirectUrl = urls_1.VaultifierUrls.getRedirectUrl();
                    // we need this secret for later OAuth token retrieval
                    storage_1.Storage.set(constants_1.StorageKey.PKCE_SECRET, pkceSecret);
                    storage_1.Storage.set(constants_1.StorageKey.OAUTH_REDIRECT_URL, redirectUrl);
                    window.location.href = vaultifier.urls.getOAuthAuthorizationCode(clientId, window.encodeURIComponent(redirectUrl), pkceSecret);
                    // we just wait forever as the browser is now changing the visible page ;-)
                    yield new Promise(() => undefined);
                }
                const idprov = oAuthType;
                if (idprov === null || idprov === void 0 ? void 0 : idprov.authority) {
                    const redirectUrl = urls_1.VaultifierUrls.getRedirectUrl();
                    storage_1.Storage.set(constants_1.StorageKey.APPLICATION_ID, idprov.applicationId);
                    storage_1.Storage.set(constants_1.StorageKey.OAUTH_REDIRECT_URL, redirectUrl);
                    const um = new oidc_client_1.UserManager({
                        authority: idprov.authority,
                        client_id: idprov.clientId,
                        scope: idprov.scope,
                        response_type: idprov.responseType,
                        redirect_uri: redirectUrl,
                    });
                    um.signinRedirect();
                    // we just wait forever as the browser is now changing the visible page ;-)
                    yield new Promise(() => undefined);
                }
            }
            const newUrl = new URL(window.location.href);
            // remove sensitive information while preserving probably important url parameters
            newUrl.searchParams.delete(appKeyParamName);
            newUrl.searchParams.delete(appSecretParamName);
            newUrl.searchParams.delete(masterKeyParamName);
            newUrl.searchParams.delete(nonceParamName);
            newUrl.searchParams.delete(clientIdParamName);
            newUrl.searchParams.delete(clientSecretParamName);
            newUrl.searchParams.delete(authorizationCodeParamName);
            newUrl.searchParams.delete(stateParamName);
            window.history.replaceState(undefined, document.title, newUrl.toString());
            return vaultifier;
        });
    }
    /**
     * Creates a Vaultifier object by retrieving connection data from URL query parameters
     */
    static create(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const getParam = VaultifierWeb._getParamAccessor();
            let _options = defaultOptions;
            if (_options)
                _options = Object.assign(Object.assign({}, _options), options);
            const { baseUrlParamName, repo, } = _options;
            let { baseUrl } = _options;
            if (!baseUrl) {
                baseUrl = getParam(baseUrlParamName);
                if (baseUrl) {
                    // in web environments we want to persist the base url
                    // if it was passed via URL parameter
                    // this is because if we use OAuth for login
                    // we'll lose all parameters after redirect, hence we have to persist it
                    storage_1.Storage.set(constants_1.StorageKey.BASE_URL, baseUrl);
                }
                else
                    // first of all we try to fetch a stored value (see above for saving the value)
                    // if this does not work we just fall back to the window's location origin, if there is no parameter specified
                    baseUrl = storage_1.Storage.get(constants_1.StorageKey.BASE_URL) || window.location.origin;
            }
            let vaultifier = new __1.Vaultifier(baseUrl, repo);
            try {
                yield vaultifier.getVaultSupport();
            }
            catch (_a) {
                // if baseUrl was specified, we try it with Vaultifier's default value
                // therefore passing undefined
                if (baseUrl) {
                    vaultifier = new __1.Vaultifier(undefined, repo);
                    try {
                        yield vaultifier.getVaultSupport();
                    }
                    catch (e) {
                        console.error(e);
                        vaultifier = undefined;
                    }
                }
            }
            return new VaultifierWeb(_options, vaultifier);
        });
    }
}
exports.VaultifierWeb = VaultifierWeb;
VaultifierWeb._getParamAccessor = () => {
    const params = new URL(window.location.href).searchParams;
    return (name) => params.get(name) || undefined;
};
