// vaultifier/dist/module/helpers.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { decrypt, isEncrypted } from './crypto';
export const parseVaultItemMeta = (data) => ({
    id: data.id,
    accessCount: data.access_count,
    createdAt: new Date(data.created_at),
    updatedAt: new Date(data.updated_at),
    tableName: data.table_name,
    dri: data.dri,
    schemaDri: data.schema_dri,
    mimeType: data.mime_type,
    merkleId: data.merkle_id,
    oydHash: data.oyd_hash,
    oydSourcePileId: data.oyd_source_pile_id,
    raw: data,
});
export const decryptOrNot = (item, privateKey) => __awaiter(void 0, void 0, void 0, function* () {
    if (privateKey &&
        isEncrypted(item)) {
        const decrypted = yield decrypt(item, { cipher: privateKey });
        try {
            return JSON.parse(decrypted);
        }
        catch ( /* the encrypted data is delivered as string */_a) { /* the encrypted data is delivered as string */ }
    }
    return item;
});
export const parseVaultItem = (data, privateKey) => __awaiter(void 0, void 0, void 0, function* () {
    if (typeof data === 'string') {
        try {
            // item usually contains JSON data, therefore we try to parse the string
            data = JSON.parse(data);
        }
        catch ( /* */_b) { /* */ }
    }
    const isContentEncrypted = isEncrypted(data.content);
    data.content = yield decryptOrNot(data.content, privateKey);
    const item = Object.assign(Object.assign({}, parseVaultItemMeta(data)), { isEncrypted: isContentEncrypted, content: data.content });
    return item;
});
