// vaultifier/dist/module/environments/web.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { Vaultifier } from "..";
import { StorageKey } from "../constants";
import { getRandomString } from '../crypto';
import { OAuthType } from "../interfaces";
import { Storage } from "../storage";
import { VaultifierUrls } from "../urls";
const defaultOptions = {
    repo: undefined,
    clientId: undefined,
    baseUrlParamName: 'PIA_URL',
    appKeyParamName: 'APP_KEY',
    appSecretParamName: 'APP_SECRET',
    masterKeyParamName: 'MASTER_KEY',
    nonceParamName: 'NONCE',
    clientIdParamName: 'client_id',
    clientSecretParamName: 'client_secret',
    authorizationCodeParamName: 'code',
};
export class VaultifierWeb {
    /**
     * Creates a Vaultifier object by retrieving connection data from URL query parameters
     */
    static create(options) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            const params = new URL(window.location.href).searchParams;
            const getParam = (name) => params.get(name) || undefined;
            let _options = defaultOptions;
            if (_options)
                _options = Object.assign(Object.assign({}, _options), options);
            const { appKeyParamName, appSecretParamName, authorizationCodeParamName, baseUrlParamName, clientIdParamName, clientSecretParamName, masterKeyParamName, nonceParamName, repo, } = _options;
            let { clientId, } = _options;
            // in web environments, we just fall back to the window's location origin, if there is no parameter specified
            const baseUrl = getParam(baseUrlParamName) || window.location.origin;
            const appKey = getParam(appKeyParamName);
            const appSecret = getParam(appSecretParamName);
            // if clientId parameter is specified as query parameter it is already the second step client id parameter of OAuth
            clientId = (_a = getParam(clientIdParamName)) !== null && _a !== void 0 ? _a : clientId;
            const authorizationCode = getParam(authorizationCodeParamName);
            let credentials = undefined;
            if (appKey && appSecret)
                credentials = {
                    appKey,
                    appSecret,
                };
            else if (authorizationCode && clientId)
                credentials = {
                    authorizationCode,
                    clientId,
                };
            const masterKey = getParam(masterKeyParamName);
            const nonce = getParam(nonceParamName);
            const end2end = (masterKey && nonce) ? {
                masterKey,
                nonce,
            } : undefined;
            let vaultifier = new Vaultifier(baseUrl, repo, credentials, end2end);
            try {
                yield vaultifier.getVaultSupport();
            }
            catch (_c) {
                // if baseUrl was specified, we try it with Vaultifier's default value
                // therefore passing undefined
                if (baseUrl) {
                    vaultifier = new Vaultifier(undefined, repo, credentials, end2end);
                    try {
                        yield vaultifier.getVaultSupport();
                    }
                    catch (e) {
                        console.error(e);
                        return undefined;
                    }
                }
            }
            try {
                // try initializing vaultifier to see if credentials are working
                yield vaultifier.initialize();
            }
            catch ( /* */_d) { /* */ }
            // if we could not authorize until this stage
            // we try to login via OAuth, if supported
            const isAuthenticated = yield vaultifier.isAuthenticated();
            if (!isAuthenticated) {
                if (clientId && ((_b = (yield vaultifier.getVaultSupport()).oAuth) === null || _b === void 0 ? void 0 : _b.type) === OAuthType.AUTHORIZATION_CODE) {
                    // create PKCE secret
                    const pkceSecret = getRandomString(32);
                    // const hashedSecret = btoa(await createSha256Hex(pkceSecret));
                    const redirectUrl = VaultifierUrls.getRedirectUrl();
                    // we need this secret for later OAuth token retrieval
                    Storage.set(StorageKey.PKCE_SECRET, pkceSecret);
                    Storage.set(StorageKey.OAUTH_REDIRECT_URL, redirectUrl);
                    window.location.href = vaultifier.urls.getOAuthAuthorizationCode(clientId, window.encodeURIComponent(redirectUrl), pkceSecret);
                    // we just wait forever as the browser is now changing the visible page ;-)
                    yield new Promise(() => undefined);
                }
            }
            const newUrl = new URL(window.location.href);
            // remove sensitive information while preserving probably important url parameters
            newUrl.searchParams.delete(appKeyParamName);
            newUrl.searchParams.delete(appSecretParamName);
            newUrl.searchParams.delete(masterKeyParamName);
            newUrl.searchParams.delete(nonceParamName);
            newUrl.searchParams.delete(clientIdParamName);
            newUrl.searchParams.delete(clientSecretParamName);
            newUrl.searchParams.delete(authorizationCodeParamName);
            window.history.replaceState(undefined, document.title, newUrl.toString());
            return vaultifier;
        });
    }
}
