// vaultifier/dist/main/communicator.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Communicator = void 0;
const axios_1 = __importDefault(require("axios"));
const errors_1 = require("./errors");
;
class Communicator {
    constructor() {
        this.setTokenCallback = (callback) => {
            this.tokenCallback = callback;
        };
        this.setNetworkAdapter = (adapter) => {
            if (adapter)
                return this.networkAdapter = adapter;
            else // default implementation
                return this.networkAdapter = {
                    get: (url, headers) => axios_1.default.get(url, {
                        headers: headers,
                    }),
                    post: (url, data, headers) => axios_1.default.post(url, data, {
                        headers: headers,
                    }),
                    put: (url, data, headers) => axios_1.default.put(url, data, {
                        headers: headers,
                    }),
                    delete: (url, headers) => axios_1.default.delete(url, {
                        headers,
                    }),
                };
        };
        this._baseHeaders = {
            'Content-Type': 'application/json',
        };
        // set default implementation
        this.networkAdapter = this.setNetworkAdapter();
    }
    _usesAuthentication() {
        return !!this.tokenCallback;
    }
    refreshToken() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.tokenCallback)
                return this.token = yield this.tokenCallback();
            return undefined;
        });
    }
    hasToken() {
        return !!this.token;
    }
    getToken() {
        return this.token;
    }
    get(url, usesAuth = false) {
        return __awaiter(this, void 0, void 0, function* () {
            return this._placeNetworkCall(() => __awaiter(this, void 0, void 0, function* () { return this.networkAdapter.get(url, this._getHeaders(usesAuth)); }), usesAuth);
        });
    }
    post(url, usesAuth = false, data) {
        return __awaiter(this, void 0, void 0, function* () {
            return this._placeNetworkCall(() => __awaiter(this, void 0, void 0, function* () { return this.networkAdapter.post(url, data, this._getHeaders(usesAuth)); }), usesAuth);
        });
    }
    put(url, usesAuth = false, data) {
        return __awaiter(this, void 0, void 0, function* () {
            return this._placeNetworkCall(() => __awaiter(this, void 0, void 0, function* () { return this.networkAdapter.put(url, data, this._getHeaders(usesAuth)); }), usesAuth);
        });
    }
    delete(url, usesAuth = false) {
        return __awaiter(this, void 0, void 0, function* () {
            return this._placeNetworkCall(() => __awaiter(this, void 0, void 0, function* () { return this.networkAdapter.delete(url, this._getHeaders(usesAuth)); }), usesAuth);
        });
    }
    tryCatch(callable) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const response = yield callable();
                return {
                    response,
                };
            }
            catch (e) {
                return {
                    error: e,
                    response: e.response,
                };
            }
        });
    }
    _placeNetworkCall(callable, isAuthenticated = false, recursionCount = 0) {
        return __awaiter(this, void 0, void 0, function* () {
            let nro = yield this.tryCatch(callable);
            if (!nro.response)
                throw nro.error ? nro.error : new Error('No network response');
            // only try to refresh authentication, if recursion count is still 0
            // otherwise we'll end up in an infinite loop
            if (isAuthenticated && recursionCount === 0) {
                // if data vault responds with a 401, our token is expired
                // therefore we fetch a new one and give the call another try
                if (nro.response.status === 401 && this._usesAuthentication()) {
                    this.token = yield this.refreshToken();
                    nro = yield this.tryCatch(callable);
                    return this._placeNetworkCall(callable, isAuthenticated, recursionCount + 1);
                }
            }
            if (nro.response.status === 401) {
                throw new errors_1.UnauthorizedError();
            }
            else if (nro.response.status >= 400) {
                throw nro.error;
            }
            return nro.response;
        });
    }
    _getHeaders(usesAuth = false) {
        return usesAuth && this._usesAuthentication() ?
            this._getDataHeaders() :
            this._baseHeaders;
    }
    _getDataHeaders() {
        if (this.token === undefined)
            throw new Error('There is no token available. Did you forget to initalize vaultifier?');
        return Object.assign(Object.assign({}, this._baseHeaders), { Accept: '*/*', Authorization: `Bearer ${this.token}` });
    }
}
exports.Communicator = Communicator;
