// vaultifier/dist/main/crypto.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateHashlink = exports.getRandomString = exports.isEncrypted = exports.decrypt = exports.encrypt = exports.createSha256Hex = exports.canonicalize = void 0;
const canonicalize_1 = __importDefault(require("canonicalize"));
// @ts-expect-error hashlink has no types
const hashlink_1 = require("hashlink");
const libsodium_wrappers_1 = __importStar(require("libsodium-wrappers"));
const core_utils_1 = require("./utils/core-utils");
exports.canonicalize = canonicalize_1.default;
exports.createSha256Hex = (value) => __awaiter(void 0, void 0, void 0, function* () {
    // browser environment
    if (typeof globalThis.crypto !== 'undefined') {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        const msgBuffer = new TextEncoder('utf-8').encode(value);
        const hashBuffer = yield crypto.subtle.digest('SHA-256', msgBuffer);
        return libsodium_wrappers_1.to_hex(new Uint8Array(hashBuffer));
    }
    // node environment
    else {
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const crypto = require('crypto');
        return crypto.createHash('sha256')
            .update(value)
            .digest('hex');
    }
});
const cryptoVersion = '0.4';
const sharedSecret = 'auth';
// the hash is created globally, so we don't use computing power to recreate it over and over
const sharedSecretHash = exports.createSha256Hex(sharedSecret);
exports.encrypt = (text, publicKey) => __awaiter(void 0, void 0, void 0, function* () {
    const nonce = libsodium_wrappers_1.default.randombytes_buf(libsodium_wrappers_1.default.crypto_box_NONCEBYTES);
    const cipherMsg = libsodium_wrappers_1.default.crypto_box_easy(libsodium_wrappers_1.from_string(text), nonce, libsodium_wrappers_1.from_hex(publicKey), libsodium_wrappers_1.from_hex(yield sharedSecretHash));
    return {
        value: libsodium_wrappers_1.to_hex(cipherMsg),
        nonce: libsodium_wrappers_1.to_hex(nonce),
        version: cryptoVersion,
    };
});
// SEE: https://libsodium.gitbook.io/doc/public-key_cryptography/authenticated_encryption
exports.decrypt = (cryptoObject, cipherObject) => __awaiter(void 0, void 0, void 0, function* () {
    const { value, nonce, version } = cryptoObject;
    const { cipher } = cipherObject;
    let { isHashed } = cipherObject;
    if (isHashed === undefined)
        isHashed = false;
    if (!!version && version !== cryptoVersion)
        throw new Error(`The provided crypto version (${version}) does not match our internal crypto version (${cryptoVersion})`);
    const _text = libsodium_wrappers_1.from_hex(value);
    const _nonce = libsodium_wrappers_1.from_hex(nonce);
    const _privKey = libsodium_wrappers_1.from_hex(yield sharedSecretHash);
    // calculates the our private key's public key
    const _pubKey = libsodium_wrappers_1.default.crypto_scalarmult_base(_privKey);
    const _cipher = libsodium_wrappers_1.from_hex(isHashed ? cipher : yield exports.createSha256Hex(cipher));
    const decrypted = libsodium_wrappers_1.default.crypto_box_open_easy(_text, _nonce, _pubKey, _cipher);
    return libsodium_wrappers_1.to_string(decrypted);
});
exports.isEncrypted = (item) => {
    return !!(item.value &&
        core_utils_1.onlyContainsHex(item.value) &&
        item.nonce &&
        core_utils_1.onlyContainsHex(item.nonce) &&
        item.version);
};
const alpha = 'abcdefghijklmnopqrstuvwxyz';
const numeric = '1234567890';
const allowedChars = `${alpha}${alpha.toUpperCase()}${numeric}`;
exports.getRandomString = (size) => {
    const arr = [];
    for (let i = 0; i < size; i++) {
        arr.push(allowedChars.charAt(Math.floor(Math.random() * allowedChars.length)));
    }
    return arr.join('');
};
const codecs = ['mh-sha2-256', 'mb-base58-btc'];
exports.generateHashlink = (data, urls = undefined, meta = undefined) => __awaiter(void 0, void 0, void 0, function* () {
    if (typeof data === 'object')
        data = exports.canonicalize(data);
    const hl = yield hashlink_1.encode({
        data: (new TextEncoder()).encode(data),
        urls,
        codecs,
        meta
    });
    return hl.split(':')[1];
});
