// vaultifier/dist/module/crypto.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import _canonicalize from 'canonicalize';
// @ts-expect-error hashlink has no types
import { encode } from 'hashlink';
import sodium, { from_hex, from_string, to_hex, to_string } from 'libsodium-wrappers';
import { onlyContainsHex } from './utils/core-utils';
export const canonicalize = _canonicalize;
export const createSha256Hex = (value) => __awaiter(void 0, void 0, void 0, function* () {
    // browser environment
    if (typeof globalThis.crypto !== 'undefined') {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        const msgBuffer = new TextEncoder('utf-8').encode(value);
        const hashBuffer = yield crypto.subtle.digest('SHA-256', msgBuffer);
        return to_hex(new Uint8Array(hashBuffer));
    }
    // node environment
    else {
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const crypto = require('crypto');
        return crypto.createHash('sha256')
            .update(value)
            .digest('hex');
    }
});
const cryptoVersion = '0.4';
const sharedSecret = 'auth';
// the hash is created globally, so we don't use computing power to recreate it over and over
const sharedSecretHash = createSha256Hex(sharedSecret);
export const encrypt = (text, publicKey) => __awaiter(void 0, void 0, void 0, function* () {
    const nonce = sodium.randombytes_buf(sodium.crypto_box_NONCEBYTES);
    const cipherMsg = sodium.crypto_box_easy(from_string(text), nonce, from_hex(publicKey), from_hex(yield sharedSecretHash));
    return {
        value: to_hex(cipherMsg),
        nonce: to_hex(nonce),
        version: cryptoVersion,
    };
});
// SEE: https://libsodium.gitbook.io/doc/public-key_cryptography/authenticated_encryption
export const decrypt = (cryptoObject, cipherObject) => __awaiter(void 0, void 0, void 0, function* () {
    const { value, nonce, version } = cryptoObject;
    const { cipher } = cipherObject;
    let { isHashed } = cipherObject;
    if (isHashed === undefined)
        isHashed = false;
    if (!!version && version !== cryptoVersion)
        throw new Error(`The provided crypto version (${version}) does not match our internal crypto version (${cryptoVersion})`);
    const _text = from_hex(value);
    const _nonce = from_hex(nonce);
    const _privKey = from_hex(yield sharedSecretHash);
    // calculates the our private key's public key
    const _pubKey = sodium.crypto_scalarmult_base(_privKey);
    const _cipher = from_hex(isHashed ? cipher : yield createSha256Hex(cipher));
    const decrypted = sodium.crypto_box_open_easy(_text, _nonce, _pubKey, _cipher);
    return to_string(decrypted);
});
export const isEncrypted = (item) => {
    return !!(item.value &&
        onlyContainsHex(item.value) &&
        item.nonce &&
        onlyContainsHex(item.nonce) &&
        item.version);
};
const alpha = 'abcdefghijklmnopqrstuvwxyz';
const numeric = '1234567890';
const allowedChars = `${alpha}${alpha.toUpperCase()}${numeric}`;
export const getRandomString = (size) => {
    const arr = [];
    for (let i = 0; i < size; i++) {
        arr.push(allowedChars.charAt(Math.floor(Math.random() * allowedChars.length)));
    }
    return arr.join('');
};
const codecs = ['mh-sha2-256', 'mb-base58-btc'];
export const generateHashlink = (data, urls = undefined, meta = undefined) => __awaiter(void 0, void 0, void 0, function* () {
    if (typeof data === 'object')
        data = canonicalize(data);
    const hl = yield encode({
        data: (new TextEncoder()).encode(data),
        urls,
        codecs,
        meta
    });
    return hl.split(':')[1];
});
